integration/duckdb/src/lance/duckdb/lance_reader.cc

```diff
@@ -0,0 +1,307 @@
+  assert(bind_data != nullptr);
+
+  auto state = std::make_unique<GlobalScanState>();
+  state->dataset = bind_data->dataset;
+
+  auto schema = state->dataset->schema();
+  std::vector<std::string> columns;
+  for (auto &column_id : input.column_ids) {
+    columns.emplace_back(schema->field(column_id)->name());
+  }
+
+  auto builder = GetResult(state->dataset->NewScan());
+  CheckStatus(builder->Project(columns));
+  auto scanner = GetResult(builder->Finish());
+  state->batch_generator = GetResult(scanner->ScanBatchesAsync());
+  return state;
+}
+
+/// Convert numeric array to duckdb vector.
+template <ArrowType T>
+void ToVector(const std::shared_ptr<::arrow::Array> &arr, ::duckdb::Vector *out) {
+  // TODO: dynamic_pointer_cast does not work here, IDK why.
+  auto array = std::static_pointer_cast<typename ::arrow::TypeTraits<T>::ArrayType>(arr);
+  assert(array != nullptr);
+  // TODO: How to use zero copy to move data from arrow to duckdb.
+  for (int i = 0; i < array->length(); ++i) {
+    out->SetValue(i, ::duckdb::Value::CreateValue(array->Value(i)));
+  }
+  out->SetVectorType(::duckdb::VectorType::FLAT_VECTOR);
+}
+
+/// Convert a String array into duckdb vector.
+template <>
+void ToVector<::arrow::StringType>(const std::shared_ptr<::arrow::Array> &arr,
+                                   ::duckdb::Vector *out) {
+  auto array = std::static_pointer_cast<::arrow::StringArray>(arr);
+  assert(array != nullptr);
+  // TODO: How to use zero copy to move data from arrow to duckdb.
+  for (int i = 0; i < array->length(); ++i) {
+    out->SetValue(i, std::string(array->Value(i)));
+  }
+  out->SetVectorType(::duckdb::VectorType::FLAT_VECTOR);
+}
+
+/// Convert a String array into duckdb vector.
+template <>
+void ToVector<::arrow::BinaryType>(const std::shared_ptr<::arrow::Array> &arr,
+                                   ::duckdb::Vector *out) {
+  auto array = std::static_pointer_cast<::arrow::BinaryArray>(arr);
+  assert(array != nullptr);
+  // TODO: How to use zero copy to move data from arrow to duckdb.
+  for (int i = 0; i < array->length(); ++i) {
+    auto val = array->Value(i);
+    out->SetValue(i, ::duckdb::Value::BLOB((::duckdb::data_ptr_t)val.data(), val.size()));
+  }
+  out->SetVectorType(::duckdb::VectorType::FLAT_VECTOR);
+}
+
+template <>
+void ToVector<::arrow::DictionaryType>(const std::shared_ptr<::arrow::Array> &arr,
+                                       ::duckdb::Vector *out) {
+  auto array = std::static_pointer_cast<::arrow::DictionaryArray>(arr);
+  // TODO: zero copy
+  out->SetVectorType(::duckdb::VectorType::FLAT_VECTOR);
+  auto dict_arr = std::dynamic_pointer_cast<::arrow::StringArray>(array->dictionary());
+  auto indices_arr = std::static_pointer_cast<::arrow::Int8Array>(array->indices());
+  for (int i = 0; i < indices_arr->length(); ++i) {
+    auto idx = indices_arr->Value(i);
+    out->SetValue(i, std::string(dict_arr->Value(idx)));
```

Comment:
i wonder whether this fixes the duckdb-arrow issue where list of dictionary querying doesn't work

Reply:
Yea probally, we can actually make (a hardcoded set of ) user defined types work via our plugins tho.